src/main/java/com/learning/mfscreener/utils/FIFOUnits.java

```diff
@@ -0,0 +1,247 @@
+package com.learning.mfscreener.utils;
+
+import com.learning.mfscreener.exception.GainsException;
+import com.learning.mfscreener.models.portfolio.Fund;
+import com.learning.mfscreener.models.portfolio.FundType;
+import com.learning.mfscreener.models.portfolio.TransactionType;
+import com.learning.mfscreener.models.portfolio.UserTransactionDTO;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.time.LocalDate;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FIFOUnits {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FIFOUnits.class);
+
+    private final Fund fund;
+    private final List<UserTransactionDTO> originalTransactions;
+    private final FundType fundType;
+    private final Map<LocalDate, MergedTransaction> mergedTransactions;
+    private final Deque<Transaction> transactions = new ArrayDeque<>();
+
+    private BigDecimal invested = BigDecimal.ZERO;
+    private BigDecimal balance = BigDecimal.ZERO;
+    private List<GainEntry> gains = new ArrayList<>();
+
+    public FIFOUnits(Fund fund, List<UserTransactionDTO> transactions) throws GainsException {
```

Comment:
@coderabbitai, Image you as an expert Java Developer please refactor the below code for more readability without any functional changes